File: daq/slc/apps/runcontrold/src/RCMaster.cc
#include "daq/slc/apps/runcontrold/RCMaster.h"
#include "daq/slc/apps/runcontrold/RCCommunicator.h"

#include <daq/slc/base/Debugger.h>
#include <daq/slc/base/StringUtil.h>

using namespace Belle2;

RCMaster::RCMaster(NSMNode* node,
                   RunConfig* config, RunStatus* status)
{
  _master_node = node;
  _config = config;
  if (_config != NULL) _config->setMaster(this);
  _status = status;
  _is_sending = false;
}

RCMaster::~RCMaster() throw()
{

}

void RCMaster::lock() throw()
{
  _mutex.lock();
}

void RCMaster::unlock() throw()
{
  _mutex.unlock();
}

void RCMaster::wait() throw()
{
  _cond.wait(_mutex);
}

void RCMaster::signal() throw()
{
  _cond.signal();
}

void RCMaster::addMasterCommunicator(RCCommunicator* comm)
{
  _mutex_comm.lock();
  _master_comm_v.push_back(comm);
  _mutex_comm.unlock();
}
void RCMaster::removeMasterCommunicator(RCCommunicator* comm)
{
  _mutex_comm.lock();
  _master_comm_v.remove(comm);
  _mutex_comm.unlock();
}

void RCMaster::sendMessageToMaster(const RunControlMessage& msg) throw()
{
  _mutex_comm.lock();
  for (std::list<RCCommunicator*>::iterator it = _master_comm_v.begin();
       it != _master_comm_v.end(); it++) {
    (*it)->sendMessage(msg);
  }
  _mutex_comm.unlock();
}

void RCMaster::sendStateToMaster(NSMNode* node) throw()
{
  _mutex_comm.lock();
  for (std::list<RCCommunicator*>::iterator it = _master_comm_v.begin();
       it != _master_comm_v.end(); it++) {
    (*it)->sendState(node);
  }
  _mutex_comm.unlock();
}

void RCMaster::sendDataObjectToMaster(const std::string& name,
                                      DataObject* data) throw()
{
  _mutex_comm.lock();
  for (std::list<RCCommunicator*>::iterator it = _master_comm_v.begin();
       it != _master_comm_v.end(); it++) {
    (*it)->sendDataObject(name, data);
  }
  _mutex_comm.unlock();
}

void RCMaster::sendLogToMaster(const SystemLog& log) throw()
{
  _mutex_comm.lock();
  for (std::list<RCCommunicator*>::iterator it = _master_comm_v.begin();
       it != _master_comm_v.end(); it++) {
    (*it)->sendLog(log);
  }
  _mutex_comm.unlock();
}

NSMNode* RCMaster::findNode(int id, const NSMMessage& msg) throw()
{
  NSMNode* node = getNodeByID(id);
  if (node == NULL) {
    const char* nodename = msg.getNodeName();
    if (nodename != NULL) {
      node = getNodeByName(nodename);
      if (node == NULL) {
        Belle2::debug("[DEBUG] Unexcepted node id: %d", id);
        return NULL;
      }
    }
  }
  return node;
}

NSMNode* RCMaster::getNodeByID(int id) throw()
{
  return (_node_id_m.find(id) != _node_id_m.end()) ? _node_id_m[id] : NULL;
}

NSMNode* RCMaster::getNodeByName(const std::string& name) throw()
{
  return (_node_name_m.find(name) != _node_name_m.end()) ? _node_name_m[name] : NULL;
}

void RCMaster::addNode(int id, NSMNode* node) throw()
{
  if (id >= 0 && node != NULL) {
    node->setNodeID(id);
    _node_id_m.insert(std::map<int, NSMNode*>::value_type(id, node));
    if (_node_name_m.find(node->getName()) == _node_name_m.end())
      _node_name_m.insert(std::map<std::string, NSMNode*>::value_type(node->getName(), node));
  }
}

void RCMaster::addNode(NSMNode* node) throw()
{
  if (node != NULL) {
    _node_v.push_back(node);
    if (_node_name_m.find(node->getName()) == _node_name_m.end())
      _node_name_m.insert(std::map<std::string, NSMNode*>::value_type(node->getName(), node));
  }
}

bool RCMaster::isSynchronized(NSMNode* node_in)
{
  if (node_in == NULL) return false;
  if (!node_in->isSynchronized()) return true;
  for (RCMaster::NSMNodeList::iterator it = getNSMNodes().begin();
       it != getNSMNodes().end(); it++) {
    NSMNode* node = *it;
    if (node == node_in) break;
    if (!node->isUsed()) continue;
    if (node->getState() != node_in->getState()) return false;
  }
  return true;
}

void RCMaster::setData(DataObject* data)
{
  _data = data;
  for (DataObject::ParamNameList::iterator it = data->getParamNames().begin();
       it != data->getParamNames().end(); it++) {
    DataObject::ParamInfoMap& param_m(data->getParams());
    std::string name = *it;
    if (param_m[name].type == DataObject::OBJECT) {
      NSMNode* node = new NSMNode(name, data->getObject(name));
      addNode(node);
      _status->add(node);
    } else if (param_m[name].type == DataObject::INT) {
      NSMNode* node = new NSMNode(name, new DataObject());
      addNode(node);
      _status->add(node);
      _config->addInt(name, data->getInt(name));
    }
  }
}

void RCMaster::setNodeControl(XMLElement* el)
{
  std::vector<XMLElement*> elc_v = el->getElements();
  for (size_t i = 0; i < elc_v.size(); i++) {
    XMLElement* elc = elc_v[i];
    for (RCMaster::NSMNodeList::iterator it = getNSMNodes().begin();
         it != getNSMNodes().end(); it++) {
      NSMNode* node = *it;
      if (elc->getAttribute("name") == node->getName()) {
        node->setUsed(!(Belle2::tolower(elc->getAttribute("used")) == "false"));
        node->setSynchronized(!(Belle2::tolower(elc->getAttribute("synchronized")) == "false"));
        break;
      }
    }
  }
}

File: daq/slc/extra/javalib/src/b2rc/core/RCMaster.java
package b2rc.core;

import java.util.ArrayList;
import java.util.HashMap;

import b2daq.core.DataObject;
import b2daq.core.RCConnection;
import b2daq.core.RCNode;
import b2daq.core.RCState;
import b2daq.java.io.ConfigFile;

public class RCMaster {

	private RCNode _master_node;
	private RunStatus _status;
	private RunConfig _config;
	private DataObject _data;
	private ArrayList<RCNode> _node_v = new ArrayList<RCNode>();
	private HashMap<Integer, RCNode> _node_id_m = new HashMap<Integer, RCNode>();
	private HashMap<String, RCNode> _node_name_m = new HashMap<String, RCNode>();
	private ConfigFile _config_file = new ConfigFile();
	private ArrayList<ConfigFile> _config_file_v = new ArrayList<ConfigFile>();
	// private RCDatabaseManager _dbmanager;

	public RCMaster(RCNode node, RunConfig config, RunStatus status) {
		_master_node = node;
		_config = config;
		_status = status;
	}

	public ArrayList<RCNode> getNSMNodes() {
		return _node_v;
	}

	public DataObject getData() {
		return _data;
	}

	public void setData(DataObject data) {
		_data = data;
		for (String name : data.getParamNames()) {
			HashMap<String, DataObject.ParamInfo> param_m = data.getParams();
			if (param_m.get(name).type == DataObject.OBJECT) {
				RCNode node = new RCNode(name, data.getObject(name));
				addNode(node);
				_status.add(node);
			} else if (param_m.get(name).type == DataObject.INT) {
				RCNode node = new RCNode(name, new DataObject());
				addNode(node);
				_status.add(node);
				_config.addInt(name, data.getInt(name));
			}
		}
	}

	public RCNode getNodeByID(int id) {
		return (_node_id_m.containsKey(id)) ? _node_id_m.get(id) : null;
	}

	public RCNode getNodeByName(String name) {
		return (_node_name_m.containsKey(name)) ? _node_name_m.get(name) : null;
	}

	public void addNode(RCNode node) {
		if (node != null) {
			_node_v.add(node);
			if (!_node_name_m.containsKey(node.getName()))
				_node_name_m.put(node.getName(), node);
		}
	}

	public void load() {
		for (RCNode node : getNSMNodes()) {
			node.setUsed(_config.getBool(node.getName()+"_used"));
			node.getState().copy(_status.getInt(node.getName()+"_state"));
			node.getConnection().copy(_status.getInt(node.getName()+"_connection"));
			if (node.getConnection().equals(RCConnection.OFFLINE)) {
				node.setState(RCState.UNKNOWN);
			}
		}
	}
	
	public void addNode(int id, RCNode node) {
		if (id >= 0 && node != null) {
			node.setNodeID(id);
			_node_id_m.put(id, node);
			if (!_node_name_m.containsKey(node.getName()))
				_node_name_m.put(node.getName(), node);
		}
	}

	public RCNode getNode() {
		return _master_node;
	}

	public RunStatus getStatus() {
		return _status;
	}

	public RunConfig getConfig() {
		return _config;
	}

	public ConfigFile getConfigFile() {
		return _config_file;
	}

	public void setConfigFile(ConfigFile config_file) {
		_config_file = config_file;
	}

	public ArrayList<ConfigFile> getConfigFiles() {
		return _config_file_v;
	}

	public void setConfigFiles(ArrayList<ConfigFile> config_file_v) {
		_config_file_v = config_file_v;
	}

}